CLI for creating Exotic quadratures

wwkong opened the pull request:
An example of a command line call is:
```
./tasgrid -meq -depth 5 -shift 1.0 -weightfile sinc0.txt -description "Test Exoquad" -symmetric -outfile exoquad.ct -print
```
where `sinc0.txt` contains a weight function surrogate/intepolant that has been generated beforehand (see [here](https://github.com/ORNL/TASMANIAN/files/7260218/sinc0.txt) for an example).

mkstoyanov commented:
* It doesn't help to add a link to a binary file, there is nothing that the anyone can get out of it. Even I can only say that this looks like a Tasmanian sparse grid file because I can read the first few characters. This is especially true when the user will never have to manually write such file, this is done using Tasmanian operations only.
* The output of the custom rule should always be in ascii format, that way the user can easily see the points and weights if they want and the `makeGlobalGrid()` method only works with an ascii custom tabulated file. The binary format for the custom tabulated was added so that the custom rule can be saved together with the rest of the sparse grid.
* is-symmetric is a flag, which means that you don't have to pass 0 or 1 from the command line, doing so is messy (what happens if the user passes 8 or a string or a floating point number). Assume that things are not symmetric (which is the more general case) and then have a flag that says `-symmetric` and that will switch to the symmetric mode.

mkstoyanov commented on SparseGrids/CMakeLists.txt:
```diff
@@ -111,6 +111,9 @@ endif()
 
 target_include_directories(Tasmanian_libsparsegrid PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/>)
+# the following are needed for Exotic quadrature only
+target_include_directories(Tasmanian_libsparsegrid PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../Addons>)
```
Note that this is technically not a proper way to use CMake. The correct thing to do is to move `tasgrid` into a separate folder that gets includes after the others. Thus, we can keep the hierarchy of sub-modules, e.g., Sparse Grid then DREAM then the Addons.

Don't do that now, we will do it after the release is finalized.

mkstoyanov commented on SparseGrids/tasgridWrapper.cpp:
```diff
@@ -362,6 +379,13 @@ void TasgridWrapper::createQuadrature(){
+    TasGrid::TasmanianSparseGrid weight_surrogate;
+    weight_surrogate.read(weightfilename.c_str());
+    // Depth = 2 * level - 1.
+    int level = depth % 2 == 1 ? (depth + 1) / 2 : depth / 2 + 1;
```
Why is level different from depth?

wwkong replied:
From what I understand, depth refers to (roughly) the number of points that make up the grid. For the exotic quadrature function though, level is the number of Gauss-Legendre correction nodes or the generated weighted Gauss nodes.

For example, if we wanted an exotic quadrature with ~15 nodes, we would want to generate 8 Gauss-Legendre nodes and 8 weighted Gauss nodes (using `decompose()`) and account for the fact that `0.0` might be common between these nodes. Hence, the "level" given to `getExoticQuadrature()` would be 8 instead of the desired depth of 15.

mkstoyanov replied:
Sparse Grid logic seldom goes by "number of points", we usually look at "level" or "polynomial order".
For the exotic quadrature, we should go by "level = depth".

An alternative could be that we add the criteria of polynomial order, i.e., find the level that matches the order but we can leave that off for now.

On a side note, even if you want to go by the number of points, there's no way to guarantee that some of the points for the shifted weight would not match the Gauss-Legendre points, so the formula may be wrong. If you want to specify the points exactly, you will have to change the logic of the construction and stop the find-roots process when you have enough total points. Let's not do that.

wwkong replied:
Gotcha. I changed level to be the same as depth in the latest push.

wwkong commented:
> * It doesn't help to add a link to a binary file, there is nothing that the anyone can get out of it. Even I can only say that this looks like a Tasmanian sparse grid file because I can read the first few characters. This is especially true when the user will never have to manually write such file, this is done using Tasmanian operations only.

Changed the attached file to be in ASCII format.

> * The output of the custom rule should always be in ascii format, that way the user can easily see the points and weights if they want and the `makeGlobalGrid()` method only works with an ascii custom tabulated file. The binary format for the custom tabulated was added so that the custom rule can be saved together with the rest of the sparse grid.

Changed.

> * is-symmetric is a flag, which means that you don't have to pass 0 or 1 from the command line, doing so is messy (what happens if the user passes 8 or a string or a floating point number). Assume that things are not symmetric (which is the more general case) and then have a flag that says `-symmetric` and that will switch to the symmetric mode.

Changed to use `-symmetric` as an optional flag.
